File: vercel/api/generate_bkk.js
import { createCanvas, loadImage, GlobalFonts } from '@napi-rs/canvas';
import fetch from 'node-fetch';
import path from 'path';
import { fileURLToPath } from 'url';
import { readFile } from 'fs/promises';
import { Blob } from 'buffer';
import FormData from 'form-data';

const __dirname = path.dirname(fileURLToPath(import.meta.url));

export default async function handler(req, res) {
  if (req.method !== 'POST') {
    return res.status(405).send('Method Not Allowed');
  }

  try {
    const {
      chat_id,
      tanggal,
      dibayarkan_kepada,
      uang_sejumlah,
      rincian = [],
      disetujui
    } = req.body;

    const fontPath = path.join(__dirname, 'PlaypenSans-Regular.ttf');
    GlobalFonts.registerFromPath(fontPath, 'Playpen');

    const image = await loadImage(path.join(__dirname, 'BKK Kosong.png'));
    const canvas = createCanvas(image.width, image.height);
    const ctx = canvas.getContext('2d');

    ctx.drawImage(image, 0, 0);
    ctx.font = '19px Playpen';
    ctx.fillStyle = 'black';

    ctx.fillText(tanggal, 160, 230);
    ctx.fillText(dibayarkan_kepada, 310, 260);
    ctx.fillText(uang_sejumlah, 310, 290);
    ctx.fillText(tanggal, 310, 320);

    let y = 380;
    rincian.forEach((item, i) => {
      ctx.fillText(String(i + 1), 75, y);
      ctx.fillText(item, 280, y);
      y += 40;
    });

    ctx.fillText(uang_sejumlah, 750, 520);
    ctx.fillText(uang_sejumlah, 700, 550);
    ctx.fillText(disetujui, 120, 670);

    const buffer = canvas.toBuffer('image/png');

    const botToken = process.env.BOT_TOKEN;
    const now = new Date();
    const time = `${String(now.getFullYear()).slice(2)}${(now.getMonth()+1).toString().padStart(2, '0')}${now.getDate().toString().padStart(2, '0')}-${now.getHours().toString().padStart(2, '0')}${now.getMinutes().toString().padStart(2, '0')}`;
    const caption = `BKK-${time}`;

    const telegramUrl = `https://api.telegram.org/bot${botToken}/sendPhoto`;
    const form = new FormData();
    form.append('chat_id', chat_id);
    form.append('caption', caption);
    form.append('photo', buffer, 'bkk.png');

    await fetch(telegramUrl, {
      method: 'POST',
      body: form
    });

    res.status(200).json({ status: 'sent' });
  } catch (err) {
    console.error(err);
    res.status(500).send('Internal Server Error: ' + err.message);
  }
}

File: vercel/api/generate_bkk.py
from PIL import Image, ImageDraw, ImageFont
from datetime import datetime
from io import BytesIO
import requests
import os

from flask import Flask, request, jsonify

app = Flask(__name__)

@app.route("/api/generate_bkk", methods=["POST"])
def generate_bkk():
    data = request.json
    chat_id = data.get("chat_id")

    tanggal = data.get("tanggal", "")
    kepada = data.get("dibayarkan_kepada", "")
    uang = data.get("uang_sejumlah", "")
    rincian = data.get("rincian", [])
    disetujui = data.get("disetujui", "")

    # Load gambar & font
    img = Image.open("BKK Kosong.png")
    draw = ImageDraw.Draw(img)
    font = ImageFont.truetype("PlaypenSans-Regular.ttf", 19)
    font_bold = font

    # Tanggal dan bagian atas
    draw.text((160, 230), tanggal, font=font, fill="black")
    draw.text((310, 260), kepada, font=font, fill="black")
    draw.text((310, 290), uang, font=font, fill="black")
    draw.text((310, 320), tanggal, font=font, fill="black")

    # Rincian
    y_start = 380
    for i, item in enumerate(rincian):
        y = y_start + i * 40
        draw.text((75, y), str(i+1), font=font, fill="black")
        draw.text((280, y), item, font=font, fill="black")

    # Total Jumlah (hanya di bagian bawah saja)
    draw.text((750, 520), uang, font=font, fill="black")
    draw.text((700, 550), uang, font=font, fill="black")

    # Disetujui oleh
    draw.text((120, 670), disetujui, font=font, fill="black")

    # Simpan ke buffer
    buffer = BytesIO()
    img.save(buffer, format="PNG")
    buffer.seek(0)

    # Buat caption dengan format BKK-yymmdd-HHmm
    time_str = datetime.now().strftime("%y%m%d-%H%M")
    caption = f"BKK-{time_str}"

    # Kirim gambar ke Telegram
    bot_token = os.environ.get("BOT_TOKEN")
    telegram_url = f"https://api.telegram.org/bot{bot_token}/sendPhoto"
    files = {
        'photo': ("bkk.png", buffer, "image/png")
    }
    data = {
        'chat_id': chat_id,
        'caption': caption
    }
    response = requests.post(telegram_url, data=data, files=files)
    return jsonify({"status": "sent", "telegram_status": response.status_code})

